### PR TITLE
Port new core features introduced in internal projects

### PR DESCRIPTION
**Pull request by lgathy**

### CODE REVIEW

File: src/main/java/com/doctusoft/math/ExponentialDelays.java [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+import static java.util.Objects.*;
+
+/**
+ * This utility helps create streams of exponentially increasing int or long values complying with various criteria.
+ * Assembling all those criteria is done in a builder style defined by {@link OfInt} and {@link OfLong} internal
+ * interfaces. Invoking their {@code build()} method returns a new {@link IntStream} or {@link LongStream} instance
+ * with the desired attributes.
+ */
+@Beta
+public interface ExponentialDelays {
```
Comment (ttozser) [inline src/main/java/com/doctusoft/math/ExponentialDelays.java]: Most of the time I want to maximize the amount of time spent on executing a callable.

For example: 
* I want a request to take at most 10 seconds. 
* I need to call two external systems. 
* The second call depends on the first so I can't do them concurrently. 

In this case I want to say something like: the first request should take at most 9 seconds.
The second should take at most the remaining time of the 10 seconds.

I can do something like this if I configure the ExponentialDelays properly, but it is not really convenient.
A limitOverallDelays method would make it easy.

A limitOverallExecution would be even better. For this you should generate the delay stream dynamically so you can consider the execution time of the task too and not just the delays.
Basically record the time time on build and stop the iteration if the next value would bring the overall time above the limit.

File: src/test/java/com/doctusoft/math/TestExponentialIntDelays.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    public void defaultDelays() {
+        assertFirstDelays(ints().build(), 1, 2, 4, 8, 16, 32, 64);
+    }
+    
+    @Test
+    public void delaysWithInitialValue() {
+        assertFirstDelays(intsFrom(3).build(), 3, 6, 12, 24, 48, 96);
+        assertThrows(() -> intsFrom(0), AnException.of(IllegalArgumentException.class));
+        assertThrows(() -> intsFrom(-1), AnException.of(IllegalArgumentException.class));
+        int randomNegative = -1 * new Random().nextInt(Integer.MAX_VALUE);
```
Comment (ttozser) [inline src/test/java/com/doctusoft/math/TestExponentialIntDelays.java]: I don't like randomness in the tests. They can easily result in flaky tests.
If the assertion fails a line below you won't even know which number caused the problem.

File: src/main/java/com/doctusoft/java/Retry.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            }
+        }
+    }
+    
+    protected void delay(long timeToWait) {
+        log.log(Level.INFO, () -> "Waiting " + printMillis(timeToWait) + " before retrying...");
+        try {
+            Thread.sleep(timeToWait);
+        } catch (InterruptedException ie) {
+            log.log(Level.FINE, "Interrupted", ie);
```
Comment (lgathy) [inline src/main/java/com/doctusoft/java/Retry.java]: Purpose was to log the interrupt event not to prevent the interrupt. Tried to make changes in order to avoid the latter.

File: src/main/java/com/doctusoft/java/RandomId.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        ByteBuffer buf = ByteBuffer.allocate(16);
+        ThreadLocalRandom random = ThreadLocalRandom.current();
+        buf.asLongBuffer()
+            .put(random.nextLong())
+            .put(random.nextLong())
+        ;
+        return CODER.encodeToString(buf.array());
+    }
+    
+    private static final PrecisionClock CLOCK = new PrecisionClock();
```
Comment (lgathy) [inline src/main/java/com/doctusoft/java/RandomId.java]: (I believe the intention was that I am usually not interrested in the private stuff when I open the source of a class, thus I put the internal details at the end)

File: src/test/java/com/doctusoft/math/TestExponentialIntDelays.java (hunk 2 above)
Comment (lgathy) [inline src/test/java/com/doctusoft/math/TestExponentialIntDelays.java]: This is how I wanted to cover things: there is one test with a negative constant (-1, a boundary value anyway) + a random to potentially uncover if for some mistake the boundary value works but anything else does not. The random is there to avoid any magic number I would have chosen alternatively (like -5). Now that I looked at it also felt reasonable to add one more constant case for Integer.MIN_VALUE as well.

File: src/main/java/com/doctusoft/math/ExponentialDelays.java (hunk 1 above)
Comment (lgathy) [inline src/main/java/com/doctusoft/math/ExponentialDelays.java]: I believe the first argument is only valid upon overloading the system thus braking it. I can think of use-cases where this is not the case or cannot happen realisticly (e.g. GAE deploy status check), but hitting the 0,1% error rate is rather likely with one request (e.g. reading from a GCS bucket). In such cases unrandomized delays should not make much of a difference.

However I see an opportunity to supporting randomized delays as well with another class called ExponentialRandomizedDelays.

The second suggestion is something I was thinking of originally as well but had been left out from the first implementation, mostly because it was not yet used in existing code. I also could not come up with a suitable name, which certainly is not an issue anymore since limitOverallExecution is an excellent suggestion. I will create 2 separate issues for this.

**Comment by lgathy** [comment]
@ttozser Did the small changes here, created #10 and #11 for separatable tasks

**Comment by lgathy** [comment]
@ttozser I am merging this and will do the remaining suggestions in the separated new issues

File: src/main/java/com/doctusoft/java/PrecisionClock.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    public Instant instant() {
+        return initial.plusNanos(getSystemNanos() - nanos);
+    }
+    
+    public Clock withZone(ZoneId zone) {
+        return new PrecisionClock(clock.withZone(zone));
+    }
+    
+    protected long getSystemNanos() {
+        return System.nanoTime();
```
Comment (nbali) [inline src/main/java/com/doctusoft/java/PrecisionClock.java]: @ttozser was the dynamic cpu clockspeed affecting and causing invalid nanotime bug fixed?

Comment (ttozser) [inline src/main/java/com/doctusoft/java/PrecisionClock.java]: I have no idea
